// Autogenerated by convert_tests.py and process_polyglot.py.
// Do not edit this file directly.
// The template for this file is located at:
// ../../../../../templates/Test.java
package gen;

import com.rethinkdb.RethinkDB;
import com.rethinkdb.gen.exc.*;
import com.rethinkdb.gen.ast.*;
import com.rethinkdb.ast.ReqlAst;
import com.rethinkdb.model.MapObject;
import com.rethinkdb.model.OptArgs;
import com.rethinkdb.net.Connection;
import com.rethinkdb.net.Cursor;
import junit.framework.TestCase;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertArrayEquals;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.junit.*;
import org.junit.rules.ExpectedException;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.Instant;
import java.util.stream.LongStream;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.concurrent.TimeoutException;
import java.util.regex.Pattern;
import java.util.Collections;
import java.nio.charset.StandardCharsets;

import static gen.TestingCommon.*;
import gen.TestingFramework;

public class MathLogicMod {
    Logger logger = LoggerFactory.getLogger(MathLogicMod.class);
    public static final RethinkDB r = RethinkDB.r;

    Connection<?> conn;
    public String hostname = TestingFramework.getConfig().getHostName();
    public int port = TestingFramework.getConfig().getPort();

    @Before
    public void setUp() throws Exception {
        conn = TestingFramework.createConnection();
        try {
            r.dbCreate("test").run(conn);
            r.db("test").wait_().run(conn);
        }catch (Exception e){}
    }

    @After
    public void tearDown() throws Exception {
        System.out.println("Tearing down.");
        if(!conn.isOpen()){
            conn.close();
            conn = TestingFramework.createConnection();
        }
        r.dbDrop("test").run(conn);
        r.db("rethinkdb").table("_debug_scratch").delete();
        conn.close(false);
    }

    // Autogenerated tests below

        @Test(timeout=120000)
    public void test() throws Exception {
                
        {
            // math_logic/mod.yaml #(1, 1)
            /* 1 */
            Long expected_ = 1L;
            /* (r.expr(10) % 3) */
            System.out.println("About to run #(1, 1): r.expr(10L).mod(3L)");
            Object obtained = runOrCatch(r.expr(10L).mod(3L),
                                          new OptArgs()
                                          ,conn);
            System.out.println("Finished running #(1, 1)");
            try {
                assertEquals(expected_, obtained);
            System.out.println("Finished asserting #(1, 1)");
            } catch (Throwable ae) {
                System.out.println("Whoops, got exception on #(1, 1):" + ae.toString());
                if(obtained instanceof Throwable) {
                    ae.addSuppressed((Throwable) obtained);
                }
                throw ae;
            }
        }
        
        {
            // math_logic/mod.yaml #(1, 2)
            /* 1 */
            Long expected_ = 1L;
            /* (10 % r.expr(3)) */
            System.out.println("About to run #(1, 2): r.mod(10L, r.expr(3L))");
            Object obtained = runOrCatch(r.mod(10L, r.expr(3L)),
                                          new OptArgs()
                                          ,conn);
            System.out.println("Finished running #(1, 2)");
            try {
                assertEquals(expected_, obtained);
            System.out.println("Finished asserting #(1, 2)");
            } catch (Throwable ae) {
                System.out.println("Whoops, got exception on #(1, 2):" + ae.toString());
                if(obtained instanceof Throwable) {
                    ae.addSuppressed((Throwable) obtained);
                }
                throw ae;
            }
        }
        
        {
            // math_logic/mod.yaml #(1, 3)
            /* 1 */
            Long expected_ = 1L;
            /* (r.expr(10).mod(3)) */
            System.out.println("About to run #(1, 3): r.expr(10L).mod(3L)");
            Object obtained = runOrCatch(r.expr(10L).mod(3L),
                                          new OptArgs()
                                          ,conn);
            System.out.println("Finished running #(1, 3)");
            try {
                assertEquals(expected_, obtained);
            System.out.println("Finished asserting #(1, 3)");
            } catch (Throwable ae) {
                System.out.println("Whoops, got exception on #(1, 3):" + ae.toString());
                if(obtained instanceof Throwable) {
                    ae.addSuppressed((Throwable) obtained);
                }
                throw ae;
            }
        }
        
        {
            // math_logic/mod.yaml #2
            /* -1 */
            Long expected_ = -1L;
            /* r.expr(-10) % -3 */
            System.out.println("About to run #2: r.expr(-10L).mod(-3L)");
            Object obtained = runOrCatch(r.expr(-10L).mod(-3L),
                                          new OptArgs()
                                          ,conn);
            System.out.println("Finished running #2");
            try {
                assertEquals(expected_, obtained);
            System.out.println("Finished asserting #2");
            } catch (Throwable ae) {
                System.out.println("Whoops, got exception on #2:" + ae.toString());
                if(obtained instanceof Throwable) {
                    ae.addSuppressed((Throwable) obtained);
                }
                throw ae;
            }
        }
        
        {
            // math_logic/mod.yaml #3
            /* err('ReqlQueryLogicError', 'Expected type NUMBER but found STRING.', [1]) */
            Err expected_ = err("ReqlQueryLogicError", "Expected type NUMBER but found STRING.", r.array(1L));
            /* r.expr(4) % 'a' */
            System.out.println("About to run #3: r.expr(4L).mod('a')");
            Object obtained = runOrCatch(r.expr(4L).mod("a"),
                                          new OptArgs()
                                          ,conn);
            System.out.println("Finished running #3");
            try {
                assertEquals(expected_, obtained);
            System.out.println("Finished asserting #3");
            } catch (Throwable ae) {
                System.out.println("Whoops, got exception on #3:" + ae.toString());
                if(obtained instanceof Throwable) {
                    ae.addSuppressed((Throwable) obtained);
                }
                throw ae;
            }
        }
        
        {
            // math_logic/mod.yaml #4
            /* err('ReqlQueryLogicError', 'Expected type NUMBER but found STRING.', [0]) */
            Err expected_ = err("ReqlQueryLogicError", "Expected type NUMBER but found STRING.", r.array(0L));
            /* r.expr('a') % 1 */
            System.out.println("About to run #4: r.expr('a').mod(1L)");
            Object obtained = runOrCatch(r.expr("a").mod(1L),
                                          new OptArgs()
                                          ,conn);
            System.out.println("Finished running #4");
            try {
                assertEquals(expected_, obtained);
            System.out.println("Finished asserting #4");
            } catch (Throwable ae) {
                System.out.println("Whoops, got exception on #4:" + ae.toString());
                if(obtained instanceof Throwable) {
                    ae.addSuppressed((Throwable) obtained);
                }
                throw ae;
            }
        }
        
        {
            // math_logic/mod.yaml #5
            /* err('ReqlQueryLogicError', 'Expected type NUMBER but found STRING.', [0]) */
            Err expected_ = err("ReqlQueryLogicError", "Expected type NUMBER but found STRING.", r.array(0L));
            /* r.expr('a') % 'b' */
            System.out.println("About to run #5: r.expr('a').mod('b')");
            Object obtained = runOrCatch(r.expr("a").mod("b"),
                                          new OptArgs()
                                          ,conn);
            System.out.println("Finished running #5");
            try {
                assertEquals(expected_, obtained);
            System.out.println("Finished asserting #5");
            } catch (Throwable ae) {
                System.out.println("Whoops, got exception on #5:" + ae.toString());
                if(obtained instanceof Throwable) {
                    ae.addSuppressed((Throwable) obtained);
                }
                throw ae;
            }
        }
    }
}
